cudapoa/samples/sample_cudapoa.cpp

```diff
@@ -124,22 +125,148 @@ void process_batch(Batch* batch, bool msa, bool print)
+        {
+            max_read_length = std::max(max_read_length, get_size<size_t>(seq) + 1);
+        }
+        max_poas[i]    = estimate_max_poas(BatchSize(max_read_length, windows[i].size()), banded_alignment, msa_flag);
+        max_lengths[i] = max_read_length;
+    }
+
+    // create histogram based on number of max POAs
+    size_t num_bins = 20;
+    std::vector<size_t> bins_frequency(num_bins, 0);             // represents the count of windows that fall within the corresponding range
+    std::vector<size_t> bins_max_length(num_bins, 0);            // represents the length of the window with maximum sequence length in the bin
+    std::vector<size_t> bins_num_reads(num_bins, 0);             // represents the number of reads in the window with maximum sequence length in the bin
+    std::vector<size_t> bins_ranges(num_bins, 1);                // represents maximum POAs
+    std::vector<std::vector<size_t>> bins_window_list(num_bins); // list of windows that are added to each bin
+
+    for (size_t j = 1; j < num_bins; j++)
+    {
+        bins_ranges[j] = bins_ranges[j - 1] * 2;
+    }
+
+    // go through all windows and keep track of the bin they fit
+    for (size_t i = 0; i < num_windows; i++)
     {
-        for (auto& seq : window)
+        for (size_t j = 0; j < num_bins; j++)
         {
-            max_read_length = std::max(max_read_length, get_size<int>(seq));
+            if (max_poas[i] <= bins_ranges[j] || j == num_bins - 1)
+            {
+                bins_frequency[j]++;
+                bins_window_list[j].push_back(i);
+                if (bins_max_length[j] < max_lengths[i])
+                {
+                    bins_max_length[j] = max_lengths[i];
+                    bins_num_reads[j]  = windows[i].size();
+                }
+                break;
+            }
         }
     }
 
-    batch_size = BatchSize(max_read_length, max_sequences_per_poa);
+    // a bin in range N means a batch made based on this bean, can launch up to N POAs. If the sum of bins frequency of higher ranges
+    // is smaller than N, they can all fit in batch N and no need to create extra batches.
+    // For example. consider the following:
+    //
+    // bins_ranges      1        2       4       8       16      32      64      128     256     512
+    // bins frequency   0        0       0       0       0       0       10      51      0       0
+    // bins width       0        0       0       0       0       0       5120    3604    0       0
+    //
+    // note that bin_ranges represent max POAs. This means larger bin ranges follow with smaller corresponding max lengths (windows with shorter reads)
+    // In the example above, to process 10 windows that fall within bin range 64, we need to create one batch. This batch can process up to 64 windows of
+    // max length 5120 or smaller. This means all the windows in bin range 128 can also be processed with the same batch and no need to launch an extra batch
```

Comment:
Does the resource allocation for the batch change if we're using windows from bin_range 64 and add bin_range 128?

Reply:
no, once a batch is created with a given `batch_size`, it remains the same for all the windows that are added to it, whether they could be launched with a different `batch_size` or not.